api/src/permissions/modules/process-ast/lib/inject-cases.ts

```diff
@@ -0,0 +1,87 @@
+import type { Filter, Permission } from '@directus/types';
+import type { AST, FieldNode, FunctionFieldNode, NestedCollectionNode } from '../../../../types/ast.js';
+import type { FieldKey } from '../types.js';
+import { dedupeAccess } from '../utils/dedupe-access.js';
+import { hasItemPermissions } from '../utils/has-item-permissions.js';
+
+/**
+ * Mutates passed AST
+ *
+ * @param ast - Read query AST
+ * @param permissions - Expected to be filtered down for the policies and action already
+ */
+export function injectCases(ast: AST, permissions: Permission[]) {
+	ast.cases = processChildren(ast.name, ast.children, permissions);
+}
+
+function processChildren(
+	collection: string,
+	children: (NestedCollectionNode | FieldNode | FunctionFieldNode)[],
+	permissions: Permission[],
+) {
+	const permissionsForCollection = permissions.filter((permission) => permission.collection === collection);
+
+	const rules = dedupeAccess(permissionsForCollection);
+	const cases: (Filter | null)[] = [];
+	const caseMap: Record<FieldKey, number[]> = {};
+
+	const requestedKeys = children.map(({ fieldKey }) => fieldKey);
+
+	for (const [index, { rule, fields }] of rules.entries()) {
+		// If none of the fields in the current permissions rule overlap with the actually requested
+		// fields in the AST, we can ignore this case altogether
+		if (fields.has('*') === false && Array.from(fields).every((field) => requestedKeys.includes(field) === false)) {
+			continue;
+		}
+
+		cases.push(rule);
+
+		for (const field of fields) {
+			caseMap[field] = [...(caseMap[field] ?? []), index];
+		}
+	}
```

Comment:
What does it mean if `rule` is `null` here? As far as I can tell it means that this permission does not have a permission filter, so it matches the `globalWhenCase` filter?  So effectively the question is what is the difference between `Permission.permissions == null` and `== {}`.

Right now `processChildren` returns `(Filter|null)[]`, which is an invalid type for `AST.cases`. 

So we should only add it to cases, if `rule != null` and use an independently incrementing index instead of the loop index?

```suggestion
	let index = 0;
	
	for (const { rule, fields } of rules) {
		// If none of the fields in the current permissions rule overlap with the actually requested
		// fields in the AST, we can ignore this case altogether
		if (fields.has('*') === false && Array.from(fields).every((field) => requestedKeys.includes(field) === false)) {
			continue;
		}

		if (rule === null) continue;
		
		cases.push(rule);

		for (const field of fields) {
			caseMap[field] = [...(caseMap[field] ?? []), index];
		}
		
		index++;
	}
```

Reply:
> So effectively another question is what is the difference between Permission.permissions == null and == {}.

There's no difference between these two. `null` and `{}` should be treated the same in the case of the permission rules

Reply:
> So we should only add it to cases, if rule != null and use an independently incrementing index instead of the loop index?

This is a good idea, as we can ignore any case that's null or `{}` anyways